Add new algorithm CreateBootstrapWorkspaces

GuiMacielPereira opened the pull request:
### Description of work
A new simple algorithm for creating batches of bootstrap workspaces. Requested by Vesuvio scientists, but applicable in general to any science area.

#### Summary of work
<!-- Please provide a short, high level description of the work that was done.
-->
The idea of bootstrap is to simulate sample workspaces by sampling with replacement from the input workspace. This algorithm introduces two ways of doing this: the first by sampling from a gaussian distribution at *each point* (the distribution has the mean at the point and deviation the error at the point) and the second way by resampling with repetition the spectra indices. This latter is more particular for the Vesuvio routines and does not make sense when the rows in the workspace are not somewhat interchangeable.

<!-- Why has this work been done? If there is no linked issue please provide appropriate context for this work.
#### Purpose of work
This can be removed if a github issue is referenced below
-->
This algorithm was requested by the Vesuvio instrument scientists, which use Bootstrap frequently in some of their analysis.

Fixes #xxxx. <!-- and fix #xxxx or close #xxxx xor resolves #xxxx. One line per issue fixed. -->
<!-- alternative
*There is no associated issue.*
-->

<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** Matthew Krzystyniak
-->

#### Further detail of work
<!-- Please provide a more detailed description of the work that has been undertaken.
-->
Most of the code is boilerplate for algorithms. The algorithm itself is rather straightforward. I would recommend to read the documentation page first and then to look at my implementation of it. The scientists provided the original implementation in Python (maybe I should have kept this algorithm in Python for simplicity sake, I didn't because I noticed that a very similar algorithm, `CreateMonteCarloWorkspace`, was already written in C++):

```python
def PyExec(self):
        in_ws = self.getPropertyValue('InputWorkspace')
        n_reps = int(self.getProperty('NumberOfReplicas').value)
        prefix = self.getPropertyValue('OutputPrefix')
        use_err = self.getProperty('UseErrorSampling').value

        ws = mtd[in_ws]
        if ws is None:
            raise RuntimeError(f"Workspace '{in_ws}' not found.")
        n_hist = int(ws.getNumberHistograms())

        # Cache X-data arrays
        x_data = [ws.readX(i) for i in range(n_hist)]

        for i in range(1, n_reps+1):
            out_name = f"{prefix}{i}"
            # Clone structure only
            CloneWorkspace(InputWorkspace=in_ws,
                           OutputWorkspace=out_name)
            clone = mtd[out_name]

            if use_err:
                # For each spectrum, sample Y from Gaussian(Y, E)
                for idx in range(n_hist):
                    y_orig = ws.readY(idx)
                    e_orig = ws.readE(idx)
                    # Draw from normal distribution per bin
                    y_new = np.random.normal(loc=y_orig, scale=e_orig)
                    clone.setX(idx, x_data[idx])
                    clone.setY(idx, y_new)
                    clone.setE(idx, e_orig)
            else:
                # Resample spectra indices with replacement
                indices = np.random.randint(0, n_hist, size=n_hist)
                for new_idx, src_idx in enumerate(indices):
                    y_src = ws.readY(int(src_idx))
                    e_src = ws.readE(int(src_idx))
                    clone.setX(new_idx, x_data[new_idx])
                    clone.setY(new_idx, y_src)
                    clone.setE(new_idx, e_src)

            logger.information(f"Created replica '{out_name}' (errorSampling={use_err})")
``` 

I added unit tests and checked that they properly fail with different expected results and with mutation of the `CreateBootstrapWorkspace` code.

### To test:

Code review:
- Read the documentation page and check spelling errors + formatting
- General C++ review: check for missing consts, pass by reference, code standards, etc, ...
- Code logic: check the two ways in the if/else statement and that they do what they are supposed to

Manual test:
- Copy the code in the documentation page and run it. Check that you get the same images.
- Load the workspace `MAR21335_Ei60meV`
- Run the algorithm with the following inputs:
- 
![image](https://github.com/user-attachments/assets/d3b3dcd6-2edd-48e6-8927-12f21265a4b4)
- Plot Spectrum 5 of `boot_1` and the original workspace and check that results are similar to image
- Now run the algorithm again but choose the type `SpectraSampling`
- Plot the colorfill for the original workspace and `boot_1` and check you get the following:
- 
![image](https://github.com/user-attachments/assets/60ce5871-36d0-40a7-8ff7-2dd9a2afcb28)

<!-- Instructions for testing.
There should be sufficient instructions for someone unfamiliar with the application to test - unless a specific
reviewer is requested.
If instructions for replicating the fault are contained in the linked issue then it is OK to refer back to these.
-->

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

GuiMacielPereira committed:
Add new algorithm for bootstraping workspaces

GuiMacielPereira committed:
Allow for overwriting workspaces when re-running algorithm

If user triggers the algorithm again with the same inputs, workspaces
should be overwritten

GuiMacielPereira committed:
Add some useful comments

GuiMacielPereira committed:
Capitalize input property

GuiMacielPereira committed:
Add documentation page

GuiMacielPereira committed:
Add release note

GuiMacielPereira committed:
Add skeleton for unit tests

Tests not currently working, this is the base for developing further

GuiMacielPereira committed:
Correct unit tests

GuiMacielPereira committed:
Change italics in documentation

GuiMacielPereira committed:
Rename algorithm to plural

Multiple workspaces are created instead of one.

GuiMacielPereira committed:
Change bootstrap type argument to be list of strings

Before, the type of boostrap was enforced as a checkbox (true or false)
and its much clearer to be a string from possible options
["ErrorSampling", "SpectraSampling"].
In the future, more types might be added, so this makes more sense.

GuiMacielPereira committed:
Use the Kernel's uniform and normal distributions

Mantid already comes with a normal and uniform int distribution for
reproduceability between different OSes when running unit tests

GuiMacielPereira committed:
Update docs with bootstrap type

GuiMacielPereira committed:
Update docs images

Since I updated the sampling distributions

walshmm commented on Framework/Algorithms/inc/MantidAlgorithms/CreateBootstrapWorkspaces.h:
```diff
@@ -0,0 +1,40 @@
+namespace Algorithms {
+using namespace std;
+
+/** CreateMonteCarloWorkspace : The algorithm generates a simulated workspace by sampling from the probability
```
Please update this comment with the correct algo name.

GuiMacielPereira replied:
Ah well spotted! I used the files of that algorithm to get started, hence these defects

walshmm commented on Framework/Algorithms/src/CreateBootstrapWorkspaces.cpp:
```diff
@@ -0,0 +1,144 @@
+
+  auto mustBePositiveOrEqualToZero = std::make_shared<BoundedValidator<int>>();
+  mustBePositiveOrEqualToZero->setLower(0);
+  declareProperty("Seed", 32, mustBePositiveOrEqualToZero,
```
just curious, why does this need to be a positive integer?  Could that not be handled internally somehow?

GuiMacielPereira replied:
You're right, it can be handled internally. For the MT generator, a negative seed gets converted to the modulus, so it gets transformed into a positive seed anyway. But I do agree that we should make it general enough so that we could change the generator to one that uses negative seeds, so I'll remove this validator

walshmm commented on Framework/Algorithms/src/CreateBootstrapWorkspaces.cpp:
```diff
@@ -0,0 +1,144 @@
+  auto mustBePositive = std::make_shared<BoundedValidator<int>>();
+  mustBePositive->setLower(1);
+  declareProperty("NumberOfReplicas", 100, mustBePositive,
+                  "Number of Monte Carlo events to simulate. Defaults to integral of input workspace if 0.");
```
Should this be updated to refer to the "ErrorSampling", "SpectraSampling" modes, i.e. Gaussian or sampling?

walshmm commented on Framework/Algorithms/src/CreateBootstrapWorkspaces.cpp:
```diff
@@ -0,0 +1,144 @@
+    progress.report("Creating Bootstrap Samples...");
+  }
+
+  auto alg = createChildAlgorithm("GroupWorkspaces");
+  alg->setProperty("InputWorkspaces", boot_names);
+  alg->executeAsChildAlg();
+  WorkspaceGroup_sptr outputGroup = alg->getProperty("OutputWorkspace");
+  setProperty("OutputWorkspaceGroup", outputGroup);
```
if the workspacegroup was made at the start you could drop the `boot_names` vector for [WorkspaceGroup#addWorkspace](https://docs.mantidproject.org/nightly/api/python/mantid/api/WorkspaceGroup.html#mantid.api.WorkspaceGroup.addWorkspace)

GuiMacielPereira replied:
Yeah that's much cleaner

GuiMacielPereira replied:
Actually it turned out that its trickier than it seemed, because I cannot use the algorithm without at least an input name. It's possible to do this without using the child algorithm, but in my opinion it's a worse implementation because it doesn't handle some edge cases very well (like overriding an already existing group workspace). So I decided to stick to using a vector of names and running the child algorithm to group workspaces at the end.

walshmm replied:
It doesnt necessarily need to be created via an algorithm.

```
from mantid.simpleapi import *
from mantid.api import WorkspaceGroup

ws1 = CreateSampleWorkspace()

# Create a group workspace and add to the ADS
ws_group = WorkspaceGroup()
mtd.add("group1", ws_group)

ws_group.add("ws1")
``

if the above is insufficient for your usecase, I dont think this comment should block the merge.  Feel free to resolve this thread if so.

jclarkeSTFC replied:
I think by using the child algorithm the creation of the group will appear in the workspace history, not sure if that happens when doing it manually.

walshmm commented on Framework/Algorithms/src/CreateBootstrapWorkspaces.cpp:
```diff
@@ -0,0 +1,144 @@
+  std::string prefix = getProperty("OutputPrefix");
+  Progress progress(this, 0, 1, numReplicas);
+
+  std::vector<std::string> boot_names;
```
should this be camelCase?  along with `new_index` later?

walshmm reviewed (commented):
Hello and thanks for your pr!  Left some comments on the code.  I'll build and run your tests, though it may take a little bit on my end depending on how out of date my local build is.

walshmm commented:
I can recreate the examples in the docs:
![image](https://github.com/user-attachments/assets/6017a7d1-a545-4a5e-8353-1c1c5bf54840)
![image](https://github.com/user-attachments/assets/3e754d2a-88f2-44c7-8701-9a149a6be4c0)


loaded the data and ran the other manual testing:
![image](https://github.com/user-attachments/assets/dedc336c-8159-4dbf-bccb-50e6c31003f8)
![image](https://github.com/user-attachments/assets/fd4068cd-0c9b-4dbc-800c-731e72f83a30)


Everything seems to align with the expected results.  Now this pr just has to pass ci.

GuiMacielPereira committed:
Fix Clang warning

GuiMacielPereira committed:
Corrected comment in header file

GuiMacielPereira committed:
Remove unecessary validator

GuiMacielPereira committed:
Correct doc strings for args

GuiMacielPereira committed:
Create wsGroup and add the workspaces to it

GuiMacielPereira committed:
Rename variable and correct comment

GuiMacielPereira committed:
Remove prefix input and automate it from ws group name

Including one argument for prefix and another for the group name did not
make sense because there was nothing to be gained from the group name
being different from the prefixes.
On the other hand, this could easily generate confusion and was prone to
errors where the algorithm was run again with the same prefix but
different group name or vice versa.

Also the decision to store the workspace names and create a group at the
end is a conscious one. Enforcing the creation of a group through the
child algorithm is preferred because it prevents bugs (for example
groups not being correctly overridden). This child algorithm cannot be
initialized without workspaces, hence the decision to use a vector of
strings to store the workspaces to group.

jclarkeSTFC commented on Framework/Algorithms/src/CreateBootstrapWorkspaces.cpp:
```diff
@@ -0,0 +1,136 @@
+    ADS.addOrReplace(wsName, bootWs);
+    bootNames.push_back(wsName);
+
+    for (size_t index = 0; index < bootWs->getNumberHistograms(); index++) {
```
It's okay, don't worry about it if there's not going to be a noticeable performance improvement.

jclarkeSTFC reviewed (changes requested):
Looks good, algorithm works as intended. To test the `ErrorSampling` option I was using the script below, I was thinking that if I ran enough samples, the mean should match the original spectra values. It then calculates the standard deviation of the generates workspaces and compares that to the original error. There seems to be a bulge at higher TOF values, I was expecting a uniform distribution, not sure why that it. It does get smaller with more generated workspaces.

The other thing is that I think the algorithm shouldn't be using the ADS if possible. Can it not just return the workspaces it creates? Then there would be no need for the ADS in the tests. If it's not possible then no problem.

```python
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

ws = Load('MAR11060')
ws = CropWorkspace(ws, StartWorkspaceIndex=4, EndWorkspaceIndex=4)
replicas = 200
CreateBootstrapWorkspaces(ws, NumberOfReplicas=replicas, OutputWorkspaceGroup='test')
workspace_string = ','.join([f"test_{i+1}" for i in range(replicas)])
print(workspace_string)
average = Mean(workspace_string)

for i in range(replicas):
    var_name = f"test_{i+1}"
    var_temp = mtd[var_name] - average
    var_temp = Power(var_temp, 2, OutputWorkspace=f"var_{i+1}")

var_string = ','.join([f"var_{i+1}" for i in range(replicas)])
variance = Mean(var_string)
DeleteWorkspaces(workspace_string)
DeleteWorkspaces(var_string)
error = Power(variance, 0.5)

error_diff = np.array(error.dataY(0)) - np.array(ws.dataE(0))
plt.plot(error_diff)
plt.show()

fig, axes = plt.subplots(dpi=94.0, edgecolor='#ffffff', num='average-1', subplot_kw={'projection': 'mantid'})
axes.plot(average, color='#1f77b4', label='average: spec 5', wkspIndex=0)
axes.plot(ws, color='#ff7f0e', label='ws: spec 5', wkspIndex=0)
axes.set_title('average')
axes.set_xlabel('Time-of-flight ($\\mu s$)')
axes.set_ylabel('Counts ($\\mu s$)$^{-1}$')
axes.set_ylim([-3.8726, 81.324])
legend = axes.legend(fontsize=8.0).set_draggable(True).legend

fig.show()

```

![image](https://github.com/user-attachments/assets/1d6a3727-2f68-4b42-8668-d1cc92dd9254)
![image](https://github.com/user-attachments/assets/e309b622-e82d-4ff2-84af-78d428b2544b)

jclarkeSTFC reviewed (approved):
We haven't worked out a way of not using the ADS in this case so can leave that as it is.
